[Coverage][Expansion] handle nested macros in scratch space

whentojump opened the pull request:
The problematic program is as follows:

```shell
#define pre_a 0
#define PRE(x) pre_##x

void f(void) {
    PRE(a) && 0;
}

int main(void) { return 0; }
```

in which after token concatenation (`##`), there's another nested macro `pre_a`.

Currently only the outer expansion region will be produced. ([compiler explorer link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:29,endLineNumber:8,positionColumn:29,positionLineNumber:8,selectionStartColumn:29,selectionStartLineNumber:8,startColumn:29,startLineNumber:8),source:'%23define+pre_a+0%0A%23define+PRE(x)+pre_%23%23x%0A%0Avoid+f(void)+%7B%0A++++PRE(a)+%26%26+0%3B%0A%7D%0A%0Aint+main(void)+%7B+return+0%3B+%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:51.69491525423727,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:cclang_assertions_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:2,lang:___c,libs:!(),options:'-fprofile-instr-generate+-fcoverage-mapping+-fcoverage-mcdc+-Xclang+-dump-coverage-mapping+',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(assertions+trunk)+(Editor+%231)',t:'0')),k:34.5741843594503,l:'4',m:28.903654485049834,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+clang+(trunk)',editorid:1,fontScale:14,fontUsePx:'0',j:2,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(assertions+trunk)+(Compiler+%232)',t:'0')),header:(),l:'4',m:71.09634551495017,n:'0',o:'',s:0,t:'0')),k:48.30508474576271,l:'3',n:'0',o:'',t:'0')),l:'2',m:100,n:'0',o:'',t:'0')),version:4))

```text
f:
  File 0, 4:14 -> 6:2 = #0
  Decision,File 0, 5:5 -> 5:16 = M:0, C:2
  Expansion,File 0, 5:5 -> 5:8 = #0 (Expanded file = 1)
  File 0, 5:15 -> 5:16 = #1
  Branch,File 0, 5:15 -> 5:16 = 0, 0 [2,0,0] 
  File 1, 2:16 -> 2:23 = #0
  File 2, 1:15 -> 1:16 = #0
  File 2, 1:15 -> 1:16 = #0
  Branch,File 2, 1:15 -> 1:16 = 0, 0 [1,2,0] 
```

The inner expansion region isn't produced because:

1. In the range-based for loop quoted below, each sloc is processed and possibly emit a corresponding expansion region.
2. For our sloc in question, its direct parent returned by `getIncludeOrExpansionLoc()` is a `<scratch space>`, because that's how `##` is processed.

    https://github.com/llvm/llvm-project/blob/88b6186af3908c55b357858eb348b5143f21c289/clang/lib/CodeGen/CoverageMappingGen.cpp#L518-L520

3. This `<scratch space>` cannot be found in the FileID mapping so `ParentFileID` will be assigned an `std::nullopt`

    https://github.com/llvm/llvm-project/blob/88b6186af3908c55b357858eb348b5143f21c289/clang/lib/CodeGen/CoverageMappingGen.cpp#L521-L526

4. As a result this iteration of for loop finishes early and no expansion region is added for the sloc.

This problem gets worse with MC/DC: as the example shows, there's a branch from File 2 but File 2 itself is missing. This will trigger assertion failures.

The fix is more or less a workaround and takes a similar approach as #89573.

~~Depends on #89573.~~ This includes #89573. Kudos to @chapuni!
This and #89573 together fix #87000: I tested locally, both the reduced program and my original use case (fwiw, Linux kernel) can run successfully.

chapuni committed:
[MC/DC][Coverage] Workaround for `##` conditions

A synthesized identifier with `##` is emitted to `<scratch space>`.
`llvm-cov` cannot handle `<scratch space> since it doesn't have actual files.

As a workaround, peel `<scratch space>` to the actual definition if the definition is present.

This affects predefined built-in macros, like __LINE__.

Fixes #87000

whentojump committed:
[Coverage][Expansion] handle nested macros in scratch space

chapuni commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -298,6 +298,22 @@ class CoverageMappingBuilder {
 
+  /// Find out where the current file is included or macro is expanded. If the
+  /// found expansion is a <scratch space>, keep looking.
+  SourceLocation getIncludeOrNonScratchExpansionLoc(SourceLocation Loc) {
```
This may be integrated to `getIncludeOrExpansionLoc(FoldScratch=true)` as an optional behavior.

chapuni commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -339,8 +355,18 @@ class CoverageMappingBuilder {
+    for (auto &Region : SourceRegions) {
       SourceLocation Loc = Region.getBeginLoc();
+
+      // Replace Region with its definition if it is in <scratch space>.
+      while (Loc.isMacroID() &&
+             SM.isWrittenInScratchSpace(SM.getSpellingLoc(Loc))) {
+        auto ExpansionRange = SM.getImmediateExpansionRange(Loc);
+        Loc = ExpansionRange.getBegin();
+        Region.setStartLoc(Loc);
+        Region.setEndLoc(ExpansionRange.getEnd());
+      }
```
Could we split out and unify this?

whentojump replied:
The common code is now being reused. But it looks a bit awkward. 
I'd like to hear your suggestions.

chapuni reviewed (commented):
I've confirmed this works. I'd like to wait for other opinions.

My comments are just my preferences.

whentojump committed:
address comments: one getIncludeOrExpansionLoc() function controlled by a flag

whentojump committed:
address comments: extract common code

whentojump committed:
address comments: return early

chapuni commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -339,8 +365,17 @@ class CoverageMappingBuilder {
+
+      // Replace Region with its definition if it is in <scratch space>.
+      auto NonScratchExpansionLoc = getNonScratchExpansionLoc(Loc);
+      Loc = NonScratchExpansionLoc.first;
+      auto EndLoc = NonScratchExpansionLoc.second;
+      Region.setStartLoc(Loc);
+      Region.setEndLoc(EndLoc.has_value() ? EndLoc.value()
+                                          : Region.getEndLoc());
```
Could you make updates optional?

```
if (has_value) {
  Loc = ...;
  Region.set(...);
}
```

chapuni commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -292,10 +292,36 @@ class CoverageMappingBuilder {
+  /// the end sloc of the found result. Otherwise the while loop didn't get
+  /// executed, which means the location wasn't changed and the caller has to
+  /// learn the end sloc from somewhere else.
+  std::pair<SourceLocation, std::optional<SourceLocation>>
```
Could it return `optional<SourceRange>`?

chapuni reviewed (commented):
They could be simpler if `SourceRegion` were compatible to `std::pair<Loc,Loc>`.
(I don't require you to work)

whentojump committed:
address comments: return optional<SourceRange>; update region only when needed

chapuni commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -292,10 +292,36 @@ class CoverageMappingBuilder {
+      EndLoc = ExpansionRange.getEnd();
+    }
+    if (EndLoc.has_value())
+      return SourceRange(Loc, EndLoc.value());
```
I wrongly supposed `isWrittenInScratchSpace()` would return `SourceRange`. Actually returns `CharSourceRange`.

You may rewind them with your previous change, if you prefer. The current implementation is not clean a bit.

whentojump replied:
> Actually returns CharSourceRange.

True. Ideally we can circulate around `CharSourceRange` without converting back and forth too much but I found it hard to make both two callers clean and straightforward. 

> You may rewind them 

I can do that, meanwhile I'll also seek if there's a better way

chapuni replied:
I meant;
```
  auto getNonScratchExpansionLoc(SourceLocation Loc) {
    std::optional<CharSourceRange> ExpansionRange = std::nullopt;
    while (Loc.isMacroID() &&
           SM.isWrittenInScratchSpace(SM.getSpellingLoc(Loc))) {
      ExpansionRange = SM.getImmediateExpansionRange(Loc);
      Loc = ExpansionRange->getBegin();
    }
    return ExpansionRange;
  }
```
(No need to follow)

ornata commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -517,7 +552,7 @@ class CoverageMappingBuilder {
     for (const auto &FM : FileIDMapping) {
       SourceLocation ExpandedLoc = FM.second.second;
-      SourceLocation ParentLoc = getIncludeOrExpansionLoc(ExpandedLoc);
+      SourceLocation ParentLoc = getIncludeOrExpansionLoc(ExpandedLoc, false);
```
The boolean parameter for this function is only used in this single callsite. Do we need it at all?

ornata replied:
I thought it was only used in one place, oops. :) The current way is fine.

chapuni commented:
@whentojump Any updates? Or Is it better for me to take this over?

whentojump committed:
Revert "address comments: return optional<SourceRange>; update region only when needed"

This reverts commit 0eeb0d3e392b1ea1ae47bc8c977eb709cd1c264c.

whentojump committed:
update region only when needed

whentojump commented:
Sorry for getting back this late.

I reverted the latest commit (using `SourceRange`) and addressed the earlier comment about updating region sloc only when they are changed.

And yes if you have better ideas about `getNonScratchExpansion` implementation you can please post a new PR.

chapuni commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -339,8 +365,18 @@ class CoverageMappingBuilder {
+    for (auto &Region : SourceRegions) {
       SourceLocation Loc = Region.getBeginLoc();
+
+      // Replace Region with its definition if it is in <scratch space>.
```
Now this collides to #91446 . Could you update?
I suggest this may come in advance of #91446.

chapuni commented on clang/lib/CodeGen/CoverageMappingGen.cpp:
```diff
@@ -339,8 +365,18 @@ class CoverageMappingBuilder {
+
+      // Replace Region with its definition if it is in <scratch space>.
+      auto NonScratchExpansionLoc = getNonScratchExpansionLoc(Loc);
+      Loc = NonScratchExpansionLoc.first;
```
I think updating Loc should be sunk into `then` clause explicitly, even if Loc is immutable when !EndLoc.

chapuni commented on clang/test/CoverageMapping/mcdc-scratch-space.c:
```diff
@@ -0,0 +1,65 @@
+// RUN: %clang_cc1 -triple %itanium_abi_triple -std=c99 -fcoverage-mcdc -fprofile-instrument=clang -fcoverage-mapping -dump-coverage-mapping -emit-llvm-only %s | FileCheck %s
```
You may overwrite 896bceb with this.

whentojump committed:
address comments: sink updating Loc should into then clause

chapuni reviewed (approved):
I'd like other approvals by clang guys. @AaronBallman Could you nominate anyone?

hanickadot reviewed (approved):
lgtm

whentojump commented:
Thanks everyone!

I notice that #87000 has been tagged with release milestone. I can handle the backport there. Maybe in a few days? Or sooner if that's needed.

chapuni commented:
@whentojump AFAIK, both of them missed boarding 18.1.6. I'll wait for opportunities.
